Windows builds, both CLI and GUI, now failing for unclear reason

Here's an example:

https://github.com/MesserLab/SLiM/actions/runs/25653704791

The error, for both builds, looks like:

> Linking CXX executable slim.exe
> D:/a/_temp/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/16.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/slim.dir/objects.a(main.cpp.obj):main.cpp:(.text.startup+0x595): undefined reference to `std::basic_filebuf<char, std::char_traits<char> >::_close()'
> collect2.exe: error: ld returned 1 exit status

So it's a link-time error, not a compile-time error.  There are no calls to close() in main.cpp for `slim`, I don't think.  There are two calls to fclose(), but the same two calls to fclose() also occur for main.cpp for `eidostool`, and that built and linked successfully, as can be seen a bit higher in the logs.  So maybe those fclose() calls are just a red herring; or maybe something about the way main.cpp for `eidostool` does things is subtly different, such that the error is not triggered.

Since it emerged spontaneously (I don't think I was working on SLiM at all at the time, it failed in the weekly test build), my guess is that GitHub changed something in their build environment for Windows.  Beyond that I have no idea.  Maybe someone here with Windows knowledge can figure it out?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -176,29 +176,6 @@ endif()
 if(WIN32)
     set(CMAKE_CXX_STANDARD_LIBRARIES "-static-libgcc -static-libstdc++ -lwsock32 -lws2_32 ${CMAKE_CXX_STANDARD_LIBRARIES}")
     set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-Bstatic,--whole-archive -lwinpthread -Wl,--no-whole-archive")
-    set(GNULIB_NAMESPACE_SOURCES "${PROJECT_SOURCE_DIR}/core/chromosome.cpp" 
-        "${PROJECT_SOURCE_DIR}/core/haplosome.cpp" 
-        "${PROJECT_SOURCE_DIR}/core/individual.cpp" 
-        "${PROJECT_SOURCE_DIR}/core/population.cpp"
-        "${PROJECT_SOURCE_DIR}/core/slim_globals.cpp"
-        "${PROJECT_SOURCE_DIR}/core/slim_sim_eidos.cpp"
-        "${PROJECT_SOURCE_DIR}/core/slim_sim.cpp"
-        "${PROJECT_SOURCE_DIR}/core/species.cpp"
-        "${PROJECT_SOURCE_DIR}/core/species_eidos.cpp"
-        "${PROJECT_SOURCE_DIR}/core/subpopulation.cpp"
-        "${PROJECT_SOURCE_DIR}/eidos/eidos_functions_files.cpp"
-        "${PROJECT_SOURCE_DIR}/eidos/eidos_functions_other.cpp"
-        "${PROJECT_SOURCE_DIR}/eidos/eidos_globals.cpp"
-        "${PROJECT_SOURCE_DIR}/eidos/eidos_class_DataFrame.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMAbout.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMAppDelegate.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMFindRecipe.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMGraphView.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMHaplotypeOptions.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMHelpWindow.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMScriptTextEdit.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMVariableBrowser.cpp"
-        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMWindow.cpp")
 endif()
 
 set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -g -Og -DDEBUG=1")
@@ -418,8 +395,7 @@ if(PARALLEL)
 	#target_link_libraries(${TARGET_NAME_SLIM} PUBLIC omp)
 endif()
 if(WIN32)
-    set_source_files_properties(${SLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h")
-    set_source_files_properties(${GNULIB_NAMESPACE_SOURCES} TARGET_DIRECTORY slim PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
+    set_source_files_properties(${SLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
     target_include_directories(${TARGET_NAME_SLIM} BEFORE PUBLIC ${GNU_DIR})
     target_link_libraries(${TARGET_NAME_SLIM} PUBLIC gnu PRIVATE bcrypt)
 endif()
@@ -440,8 +416,7 @@ if(PARALLEL)
 	#target_link_libraries(${TARGET_NAME_EIDOS} PUBLIC omp)
 endif()
 if(WIN32)
-    set_source_files_properties(${EIDOS_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h")
-    set_source_files_properties(${GNULIB_NAMESPACE_SOURCES} TARGET_DIRECTORY slim eidos PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
+    set_source_files_properties(${EIDOS_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
     target_include_directories(${TARGET_NAME_EIDOS} BEFORE PUBLIC ${GNU_DIR})
     target_link_libraries(${TARGET_NAME_EIDOS} PUBLIC gnu PRIVATE bcrypt)
 endif()
@@ -514,8 +489,7 @@ if(BUILD_SLIMGUI)
     target_link_libraries( ${TARGET_NAME_SLIMGUI} PUBLIC OpenGL::GL gsl tables eidos_zlib )
   else()
     if(WIN32)
-      set_source_files_properties(${QTSLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h")
-      set_source_files_properties(${GNULIB_NAMESPACE_SOURCES} TARGET_DIRECTORY slim eidos SLiMgui PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
+      set_source_files_properties(${QTSLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
       target_include_directories(${TARGET_NAME_SLIMGUI} BEFORE PUBLIC ${GNU_DIR})
       target_link_libraries(${TARGET_NAME_SLIMGUI} PUBLIC OpenGL::GL gsl tables eidos_zlib gnu PRIVATE bcrypt)
     else()

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -176,6 +176,30 @@ endif()
 if(WIN32)
     set(CMAKE_CXX_STANDARD_LIBRARIES "-static-libgcc -static-libstdc++ -lwsock32 -lws2_32 ${CMAKE_CXX_STANDARD_LIBRARIES}")
     set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-Bstatic,--whole-archive -lwinpthread -Wl,--no-whole-archive")
+    set(GNULIB_NAMESPACE_SOURCES "${PROJECT_SOURCE_DIR}/core/chromosome.cpp"
+        "${PROJECT_SOURCE_DIR}/core/haplosome.cpp"
+        "${PROJECT_SOURCE_DIR}/core/individual.cpp"
+        "${PROJECT_SOURCE_DIR}/core/main.cpp"
+        "${PROJECT_SOURCE_DIR}/core/population.cpp"
+        "${PROJECT_SOURCE_DIR}/core/slim_globals.cpp"
+        "${PROJECT_SOURCE_DIR}/core/slim_sim_eidos.cpp"
+        "${PROJECT_SOURCE_DIR}/core/slim_sim.cpp"
+        "${PROJECT_SOURCE_DIR}/core/species.cpp"
+        "${PROJECT_SOURCE_DIR}/core/species_eidos.cpp"
+        "${PROJECT_SOURCE_DIR}/core/subpopulation.cpp"
+        "${PROJECT_SOURCE_DIR}/eidos/eidos_functions_files.cpp"
+        "${PROJECT_SOURCE_DIR}/eidos/eidos_functions_other.cpp"
+        "${PROJECT_SOURCE_DIR}/eidos/eidos_globals.cpp"
+        "${PROJECT_SOURCE_DIR}/eidos/eidos_class_DataFrame.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMAbout.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMAppDelegate.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMFindRecipe.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMGraphView.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMHaplotypeOptions.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMHelpWindow.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMScriptTextEdit.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMVariableBrowser.cpp"
+        "${PROJECT_SOURCE_DIR}/QtSLiM/QtSLiMWindow.cpp")
 endif()
 
 set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -g -Og -DDEBUG=1")
@@ -395,7 +419,8 @@ if(PARALLEL)
 	#target_link_libraries(${TARGET_NAME_SLIM} PUBLIC omp)
 endif()
 if(WIN32)
-    set_source_files_properties(${SLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
+    set_source_files_properties(${SLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h")
+    set_source_files_properties(${GNULIB_NAMESPACE_SOURCES} TARGET_DIRECTORY slim PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
     target_include_directories(${TARGET_NAME_SLIM} BEFORE PUBLIC ${GNU_DIR})
     target_link_libraries(${TARGET_NAME_SLIM} PUBLIC gnu PRIVATE bcrypt)
 endif()
@@ -416,7 +441,8 @@ if(PARALLEL)
 	#target_link_libraries(${TARGET_NAME_EIDOS} PUBLIC omp)
 endif()
 if(WIN32)
-    set_source_files_properties(${EIDOS_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
+    set_source_files_properties(${EIDOS_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h")
+    set_source_files_properties(${GNULIB_NAMESPACE_SOURCES} TARGET_DIRECTORY slim eidos PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
     target_include_directories(${TARGET_NAME_EIDOS} BEFORE PUBLIC ${GNU_DIR})
     target_link_libraries(${TARGET_NAME_EIDOS} PUBLIC gnu PRIVATE bcrypt)
 endif()
@@ -489,7 +515,8 @@ if(BUILD_SLIMGUI)
     target_link_libraries( ${TARGET_NAME_SLIMGUI} PUBLIC OpenGL::GL gsl tables eidos_zlib )
   else()
     if(WIN32)
-      set_source_files_properties(${QTSLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
+      set_source_files_properties(${QTSLIM_SOURCES} PROPERTIES COMPILE_FLAGS "-include config.h")
+      set_source_files_properties(${GNULIB_NAMESPACE_SOURCES} TARGET_DIRECTORY slim eidos SLiMgui PROPERTIES COMPILE_FLAGS "-include config.h -DGNULIB_NAMESPACE=gnulib")
       target_include_directories(${TARGET_NAME_SLIMGUI} BEFORE PUBLIC ${GNU_DIR})
       target_link_libraries(${TARGET_NAME_SLIMGUI} PUBLIC OpenGL::GL gsl tables eidos_zlib gnu PRIVATE bcrypt)
     else()

